server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -588,6 +622,223 @@ private Scheduler.Cancellable fetchClusterFormationInfo(
+                        CoordinationDiagnosticsAction.Response::new
+                    )
+                );
+            }
+        }, e -> {
+            logger.warn("Exception connecting to master node", e);
+            responseConsumer.accept(new RemoteMasterHealthResult(node, null, e));
+        });
+
+        fetchCoordinationDiagnosticsListener.whenComplete(response -> {
+            long endTime = System.nanoTime();
+            logger.trace("Received master stability result from {} in {}", node, TimeValue.timeValueNanos(endTime - startTime));
+            /*
+             * It is possible that the task has been cancelled at this point, but it does not really matter. In that case this result
+             * will be ignored and soon garbage collected.
+             */
+            responseConsumer.accept(new RemoteMasterHealthResult(node, response.getCoordinationDiagnosticsResult(), null));
+        }, e -> {
+            logger.warn("Exception in master stability request to master node", e);
+            responseConsumer.accept(new RemoteMasterHealthResult(node, null, e));
+        });
+
+        return transportService.getThreadPool().schedule(() -> {
+            Version minSupportedVersion = Version.V_8_4_0;
+            if (node.getVersion().onOrAfter(minSupportedVersion) == false) { // This was introduced in 8.4.0
+                logger.trace(
+                    "Cannot get cluster coordination info for {} because it is at version {} and {} is required",
+                    node,
+                    node.getVersion(),
+                    minSupportedVersion
+                );
+            } else if (isCancelled.get()) {
+                logger.trace("The remote coordination diagnostics task has been cancelled, so not opening a remote connection");
+            } else {
+                /*
+                 * Since this block is not synchronized it is possible that this task is cancelled between the check above and when the
+                 * code below is run, but this is harmless and not worth the additional synchronization in the normal case. In that case
+                 * the connection will just be closed and the transport request will not be made.
+                 */
+                transportService.connectToNode(
+                    // Note: This connection must be explicitly closed in the connectionListener
+                    node,
+                    ConnectionProfile.buildDefaultConnectionProfile(clusterService.getSettings()),
+                    connectionListener
+                );
+            }
+        }, new TimeValue(10, TimeUnit.SECONDS), ThreadPool.Names.SAME);
+    }
+
+    private void cancelPollingRemoteStableMasterHealthIndicatorService() {
+        synchronized (remoteDiagnosticsMutex) {
+            if (remoteStableMasterHealthIndicatorTask != null && remoteCoordinationDiagnosticsCancelled.getAndSet(true) == false) {
```

Comment:
`remoteCoordinationDiagnosticsCancelled.getAndSet(true) == false`
Is `compareAndSwap` a better option here?

Reply:
I don't think so, for the same reasons given above.